server/src/main/java/org/elasticsearch/cluster/node/NodeVersions.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.node;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.index.IndexVersion;
+
+import java.util.Objects;
+
+/**
+ * Represents the versions of various aspects of an Elasticsearch node.
+ * @param nodeVersion   The release version
+ * @param minIndexVersion   The minimum {@link IndexVersion} supported by this node
+ * @param maxIndexVersion   The maximum {@link IndexVersion} supported by this node
+ */
+public record NodeVersions(Version nodeVersion, IndexVersion minIndexVersion, IndexVersion maxIndexVersion) {
+
+    public static final NodeVersions CURRENT = new NodeVersions(Version.CURRENT, IndexVersion.MINIMUM_COMPATIBLE, IndexVersion.CURRENT);
```

Comment:
I anticipate we will need a similar mechanism for serverless overriding index version (for index behavior changes, not lucene). We should probably adjust IndexVersion.CURRENT to be a method like we have now in TransportVersion. I think we will need to similarly lazily init current here, otherwise it will force IndexVersion.CURRENT to be loaded even if it is moved to a method.

Reply:
I'll do that in a separate PR

Reply:
https://github.com/elastic/elasticsearch/pull/97132